addons/project/controllers/public_project.py

```diff
@@ -0,0 +1,369 @@
+            if project.check_access_rights('write'):
+                user_rights = 2  # edit rights
+            elif project.check_access_rights('read'):
+                user_rights = 1  # readonly rights
+        if self._public_project_check_token(access_token, project_id=project_id, task_id=task_id, action='edit'):
+            user_rights = 2  # edit rights
+        elif self._public_project_check_token(access_token, project_id=project_id, task_id=task_id, action='read'):
+            user_rights = max(1, user_rights)  # readonly rights, unless the logged user has higher rights
+        if action == 'read':
+            return user_rights >= 1
+        if action == 'edit':
+            return user_rights == 2
+        return levels[user_rights]
+
+    def _chatter_get_track_args(self, task_id, fields_to_track):
+        """Get the values required by task.message_track, as well as the task itself.
+
+        Note: Must be called before the write operation in order to get initial values.
+
+        :param task_id: (int) The ID of the task to track
+        :param fields_to_track: ([str]) The list of fields to track on the task
+        :return: (Object, dict, dict) The task object, tracked fields and their initial values
+        """
+        task = request.env['project.task']
+        task = task.sudo() if not self._is_logged_in() else task
+        task = task.browse(task_id) if task_id else task
+        tracked_fields = task.fields_get(fields_to_track) if task else None
+        initial_values = dict([(task.id, {field: task[field] for field in fields_to_track})])
+        return task, tracked_fields, initial_values
+
+    def _chatter_track(self, task, tracked_fields, initial_values):
+        """Track fields of a task for the chatter, as Public User if user is not logged in.
+
+        :param task: (Object) The task object
+        :param tracked_fields: (dict) The tracked fields
+        :param initial_values: (dict) The tracked fields' initial values
+        :return: True
+        """
+        kwargs = {'author_id': request.env.uid} if not self._is_logged_in() else {}
+        task.message_track(tracked_fields, initial_values, **kwargs)
+        return True
+
+    def _check_and_perform(self, method_to_call, submethod=None, access_info=None, **kw):
+        """Check permissions before calling a given method.
+
+        :param method_to_call: (str) The name of the method to call if access is granted.
+        :param submethod: (str) The name of the method that the method to call will call
+                                if access is granted (if method_to_call == call or call_kw).
+                                Defaults to None.
+        :param kw: (dict) All args and kwargs passed to the method to call.
+        :return: (any) The result of the method to call or None if no access was granted.
+        """
+        read_words = ['read', 'load', 'default', 'get', 'search',
```

Comment:
Extract the actual list of methods (not substring/keywords) and move it to dedicated methods that could be overridden in other modules if needed. This will make sure we never allow unsafe methods to be called, or private methods, etc.